### ISSUE TITLE
Breaking change in minor version?

### ISSUE THREAD

**Issue by SHxKM** [plamere/spotipy#828]
Hi,

Super glad to see this maintained after a long halt.

I've noticed after upgrading from `2.4.4.` to `2.19.0` that some endpoints like `search` no longer work when the `Spotipy` object is initialized with a `auth` param. 

My old code (now returns an error):

```python
credentials = oauth2.SpotifyClientCredentials(
        client_id=SPOTIFY_CLIENT_ID,
        client_secret=SPOTIFY_SECRET_ID,
    )
token = credentials.get_access_token()
spotipy_obj = spotipy.Spotify(auth=token)
artist_on_spotify = spotipy_obj.search(q="X", type="artist", limit=1)
```

The new code:

```python
spotipy_obj = spotipy.Spotify(client_credentials_manager=credentials)
artist_on_spotify = spotipy_obj.search(q="X", type="artist", limit=1)
```

Meanwhile, user-related endpoints continue to work with `token`.

Is this expected? I couldn't find anything in the release notes regarding this. Maybe I didn't look hard enough.

Also: am I correct in saying that the cache system is only for the application's access token, not those of the application's users?

**Comment by Peter-Schorn** [comment]
>My old code (now returns an error):
>
>```
>credentials = >oauth2.SpotifyClientCredentials(
>        >client_id=SPOTIFY_CLIENT_ID,
>        >client_secret=SPOTIFY_SECRET_ID,
    )
>token = >credentials.get_access_token()
>spotipy_obj = >spotipy.Spotify(auth=token)
>artist_on_spotify = >spotipy_obj.search(q="X", >type="artist", limit=1)
>```

This will always return an error after an hour when the access token expires. The refresh token has not been saved.

Therefore, this pattern is _never_ recommended.

**Comment by SHxKM** [comment]
@Peter-Schorn that code is just a simplified example. I’m not complaining that it’s changed I just want to know if it indeed was.

**Comment by Peter-Schorn** [comment]
Your code is flawed for any version as already described.

**Comment by SHxKM** [comment]
> Your code is flawed for any version as already described.

You can repeat yourself being condescending all you want. The question remains: a code that works in 2.4.4 doesn’t work 2.19.0, is this breaking change intended?

**Comment by Peter-Schorn** [comment]
Don't know.

**Comment by SHxKM** [comment]
Hi @stephanebruckert, thank you. With `2.19.0` here's the exception of type `spotipy.client.SpotifyException`:

```
http status: 400, code:-1 - https://api.spotify.com/v1/search?q=Roped+Off+Problem+%26+Boogie+artist%3AThe+Game&limit=50&offset=0&type=track&market=us:
 Only valid bearer authentication supported, reason: None
```

As I said, with `2.4.4` this works fine.

**Comment by stephanebruckert** [comment]
@SHxKM in your "old" code can you try to replace

> token = credentials.get_access_token()

with 

> token = credentials.get_access_token(as_dict=False)

and let me know if it helps?

**Comment by SHxKM** [comment]
> @SHxKM in your "old" code can you try to replace
> 
> ```
> token = credentials.get_access_token()
> ```
> 
> with
> 
> ```
> token = credentials.get_access_token(as_dict=False)
> ```
> 
> and let me know if it helps?

That does not help unfortunately.

I really have no gripes or complaints. Just want to make sure _**I'm** not_ missing anything with this upgrade as I figured it shouldn't include breaking changes..

**Comment by SHxKM** [comment]
> @SHxKM in your "old" code can you try to replace
> 
> ```
> token = credentials.get_access_token()
> ```
> 
> with
> 
> ```
> token = credentials.get_access_token(as_dict=False)
> ```
> 
> and let me know if it helps?

Yes, `as_dict=False` solves the issue. Can you elaborate on this please?

**Comment by stephanebruckert** [comment]
Since 2.4.4, was added a new parameter `as_dict`.

https://github.com/plamere/spotipy/blob/06551cd0553d6e030f3f4ee7da5fac12c424bac7/spotipy/oauth2.py#L213

Looks like it should be False by default instead of True for things to remain backward-compatible

**Comment by stephanebruckert** [comment]
> By the way, is the spotipy.Spotify object designed to a be a singleton?

It looks like it is, I would try to initialise it once and pass it where you need. Particularly since the tokens auto-refresh in newer versions!

**Comment by SHxKM** [comment]
> > By the way, is the spotipy.Spotify object designed to a be a singleton?
> 
> It looks like it is, I would try to initialise it once and pass it where you need. Particularly since the tokens auto-refresh in newer versions!

Got it. I wonder if that'll mess things up when I'm authenticating both on the app-level and the-user level. But that's for another issue. Thanks a lot @stephanebruckert!

**Comment by Peter-Schorn** [comment]
> Got it. I wonder if that'll mess things up when I'm authenticating both on the app-level and the-user level. But that's for another issue.

`Spotify` is not a singleton, but you should generally only be using *one* instance of it in your program, especially if you've only authenticated with one user. Do not authenticate separately for different endpoints. That's a waste of your time.
